k256/src/schnorr.rs

```diff
@@ -0,0 +1,518 @@
+//! Taproot Schnorr signatures as defined in [BIP340].
+//!
+//! [BIP340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+
+#![allow(non_snake_case, clippy::many_single_char_names)]
+
+use crate::{
+    arithmetic::FieldElement, AffinePoint, FieldBytes, NonZeroScalar, ProjectivePoint, PublicKey,
+    Scalar,
+};
+use ecdsa_core::signature::{Error, Result};
+use elliptic_curve::{
+    bigint::U256,
+    ops::{LinearCombination, Reduce},
+    rand_core::{CryptoRng, RngCore},
+    subtle::ConditionallySelectable,
+    DecompactPoint,
+};
+use sha2::{Digest, Sha256};
+
+const AUX_TAG: &[u8] = b"BIP0340/aux";
+const NONCE_TAG: &[u8] = b"BIP0340/nonce";
+const CHALLENGE_TAG: &[u8] = b"BIP0340/challenge";
+
+/// Taproot Schnorr signature as defined in [BIP340].
+///
+/// [BIP340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct Signature {
+    // TODO(nickray): maybe store (FieldElement, NonScalar), avoiding the unwraps.
+    // However, this
+    // a) contradicts tarcieri's philosophy of storing "bag of bytes"
+    // b) makes the Debug, Eq, PartialEq, PartialOrd, Ord impls tricky,
+    // as FieldElement + NonZeroScalar don't implement them, so we'd have to
+    // either modify upstream or implement in terms of the calculated `to_bytes`.
```

Comment:
I think it's fine to store the parsed components of the signature alongside `bytes`.

To simplify some of the questions you can remove the `PartialOrd` and `Ord` derives. I'm not sure they're useful for anything.

A handwritten `Debug` would probably make sense, either displaying the byte serialization or the parsed components of the signature.